README.md
# Meghanada-Server

[![Join the chat at https://gitter.im/mopemope/meghanada-server](https://badges.gitter.im/mopemope/meghanada-server.svg)](https://gitter.im/mopemope/meghanada-server?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) 
[![Patreon](https://img.shields.io/badge/patreon-become%20a%20patron-red.svg)](https://www.patreon.com/mopemope)
[![Github](https://github.com/mopemope/meghanada-server/workflows/Java%20CI/badge.svg)](https://github.com/mopemope/meghanada-server/actions)

A Java IDE Server for your editor. Java IDE-like features to your favourite text editor.

## Features

Some planned and implemented features:

* Server supports a network connection
* `Gradle` and `Maven` and `Eclipse` project support
* Run build tool task
* Compile your project
* Support annotaion processor
* Analyze java source (hooks into build)
* Code completion
* Optimize import
* Jump declaration (without source)
* Run junit test (include test runner)
* Search references
* Full-featured text search (default off)

Meghanada-Server support only emacs client (meghanada-mode)

The Meghanada architecture is almost the same as `ensime`. It is client server model.

Meghanada updates any information when saving and compile the java file.

## Building

### Requirement

* JDK 8 or later.

If your project were maven project, It needs `maven` and add `mvn` command your `$PATH`.

### Build jar

```
./gradlew clean goJF check shadowJar
```

## Usage Server

See help.

```
java -jar path/to/meghanada.jar --help
```

```
usage: meghanada server
    --gradle-version <arg>   set use gradle version
 -h,--help                   show help
 -l,--log <arg>              log file location. default:
                             /tmp/meghanada_server.log
    --output <arg>           output format (sexp, csv, json). default:
                             sexp
 -p,--port <arg>             set server port. default: 55555
 -r,--project <arg>          set project root path. default: current path
 -v,--verbose                show verbose message (DEBUG)
    --version                show version information
 -vv,--traceVerbose          show verbose message (TRACE)
```


## Run Server

```
java -jar path/to/meghanada.jar
```

Recommend settings jvm args.

```
java -XX:+UseConcMarkSweepGC -XX:SoftRefLRUPolicyMSPerMB=50 -Xverify:none -Xms256m -Dfile.encoding=UTF-8 -jar path/to/meghanada.jar
```

Meghanada-Server is required JDK 8 or later (not JRE). It used Compiler API.

## Contributing

Contributions are extremely welcome!

Please check execute the following command before contributing. then please push PR to `dev` branch.

```
./gradlew clean goJF check
```

### Customize project manually

* Write `.meghanada.conf` on project root.

example:

```
# This is an annotated reference of the options that may be set in a
# .meghanada.conf file.
#

# Set JAVA_HOME
# Type: string
java-home = "/usr/lib/jvm/default

# Set java version
# Type: string
#
# It is same effect following code.
# System.setProperty("java.specification.version", val);
java-version = "1.8"

# Set source compatibility
# Type: string
compile-source = "1.8"

# Set target compatibility
# Type: string
compile-target = "1.8"

java11-javac-args = [
                 "--add-exports",
                 "jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED",
                 "--add-exports",
                 "jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
                 "--add-exports",
                 "jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED",
                 "--add-exports",
                 "jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
                 "--add-exports",
                 "jdk.compiler/com.sun.tools.javac.resources=ALL-UNNAMED",
                 "--add-exports",
                 "jdk.compiler/com.sun.tools.javac.resources=ALL-UNNAMED",
                 "--add-exports",
                 "java.management/sun.management=ALL-UNNAMED",
]

# Set dependencies file list (jar filepath)
# Type: string list
dependencies = ["/home/user/.m2/repository/org/apache/maven/maven-model/3.3.9/maven-model-3.3.9.jar", "/home/user/.m2/repository/org/codehaus/plexus/plexus-utils/3.0.22/plexus-utils-3.0.22.jar", "/home/user/.m2/repository/org/apache/commons/commons-lang3/3.4/commons-lang3-3.4.jar" ... ]

# Set test dependencies file list (jar filepath)
# Type: string list
test-dependencies = ["/home/ma2/.m2/repository/junit/junit/4.12/junit-4.12.jar" ... ]

# Set source directories
# Type: string list
sources = ["src/main/java"]

# Set resource directories
# Type: string list
resources = ["src/main/resources"]

# Set classes output directory
# Type: string
output = "build/main/classes"

# Set test source directories
# Type: string list
test-sources = ["src/test/java"]

# Set testt resource directories
# Type: string list
test-resources = ["src/test/resources"]

# Set test classes output directory
# Type: string
test-output = "build/test/classes"

```

## License

GPL v3, See [LICENSE](LICENSE) file.


RELEASES.md
# Version 1.2.1 (2019-12-15)

* fix some bugs and improve stability.

# Version 1.2.0 (2019-05-26)

* collect telemetry data of meghanada-server (default disabled).
* fix constructor completion format.
* fix array completion.
* fix file paths not being processed correctly in windows.
* fix some bugs and improve stability.

# Version 1.1.2 (2019-05-09)

* fix NoClassDefFoundError when run junit.
* fix some bugs and improve stability.

# Version 1.1.1 (2019-05-07)

* fix detect gradle project root.
* fix some bugs and improve stability.

# Version 1.1.0 (2019-05-06)

* support open j9 (experimental).
* fix some bugs and improve stability.

# Version 1.0.14 (2019-04-26)

* works without the project tool (mvn, gradle, etc).
* support java 12 (experimental).
* add jvm option setting (meghanada-jvm-option).
* support company-box.
* fix some bugs and improve stability.

# Version 1.0.13 (2019-02-25)

* support build.gradle.kts.
* fix some bugs and improve stability.

# Version 1.0.12 (2019-01-22)

* Support Java 11.
* Fix some bugs and improve stability.

# Version 1.0.11 (2019-01-09)

* Improve completion order.
* Support annotaion value completion.
* Fix some bugs and improve stability.

# Version 1.0.10 (2018-11-16)

* Support junit5 and junit4. (thanks @roman)
* Add extra arguments to exec/debug main commands. (thanks @roman)
* Fix some bugs and improve stability.

# Version 1.0.9 (2018-09-10)

* Add jump-symbol. (thanks @iocanel)
* Fix search for junit test.
* Display maven error details.
* Fix some bugs and improve stability.

# Version 1.0.8 (2018-07-18)

* Update libs.

# Version 1.0.7 (2018-06-28)

* Fix chained completion.

# Version 1.0.6 (2018-06-20)

* Fix broken completion.
* Fix some bugs and improve stability.

# Version 1.0.5 (2018-06-13)

* Fix broken completion.
* Fix some bugs and improve stability.

# Version 1.0.4 (2018-06-05)

* Support eclipse project (experimental).
* Fix broken custom source formatter.
* Fix broken optimize import.
* Fix decompiled source parse error.
* Decrease cpu usage  a little.
* Fix some bugs and improve stability.

# Version 1.0.3 (2018-05-24)

* Support import completion from symbol (added import-at-point function).
* Completion matcher is selectable (default is prefix matcher).
* Fix broken static import completion.
* Fix import statement sort (in case google-java-format).
* Fix some bugs and improve stability.

# Version 1.0.2 (2018-05-08)

* Show progress when meghanada downloading server module
* Fix some bugs and improve stability.

# Version 1.0.1 (2018-04-26)

* Disable full-text-search by default
* Supported import static method completion (experimental)
* Fix package completion
* Fix some bugs and improve stability.

# Version 1.0.0 (2018-04-07)

## Highlights

* Change to download server module from setup program
* Add full-featured text search as search-everywhere (meghanada-searcheverywhere)
* Supported to automatically assign port number
* Decrease jar file size
* Fix bugs and improve stability.

# Version 0.9.2 (2018-02-22)

* Fix bugs and improve stability.
* Update some libraries

# Version 0.9.1 (2018-02-14)

* Initial support for Windows
* Experimental support Android plugin for gradle. (supported only 3.0.0+)
* Update some libraries
* Fix bugs and improve stability.

# Version 0.9.0 (2017-12-31)

* Initial support for Java 9
* Changeable meghanada cache directory
* Update some libraries
* Fix bugs and improve stability.

# Version 0.8.4 (2017-11-29)

* Update some libraries
* Fix bugs and improve stability

# Version 0.8.3 (2017-07-25)

* Support external debugger
* Fix autocompletion when use multiline statement
* Fix bugs and improve stability.

# Version 0.8.2 (2017-06-29)

## Highlights

* Add type information command (meghanada-typeinfo)

# Version 0.8.1 (2017-06-26)

* Fix non-escaped code string used for reference.
* Fix implicit type conversion bugs

# Version 0.8.0 (2017-06-26)

## Highlights

* Support main class execution (meghanada-exec-main).
* Support on the fly syntax checking (and analyze, compile) .
* Support search reference.
* Fix bugs and improve stability.

# Version 0.7.13 (2017-06-12)

* Fix cache update bugs

# Version 0.7.11 (2017-06-09)

* Change to use fast-serialization.
* Change to use xodous for cache backend.
* Change not to build gradle subproject by default
* Fix auto change project.
* Fix jump to references to super class's field.
* Fix optimize import is broken.
* Fix bugs and improve stability.

# Version 0.7.10 (2017-05-26)

* Fix incorrect compile target file.
* Fix import completion.
* Fix compilation sort.
* Fix bugs and improve stability.

# Version 0.7.9 (2017-05-24)

* Change completion sort to a better sort.
* Fix Sexp parse error.
* Fix crash when displaying eldoc.
* Fix an empty line on optimize import.
* Fix many bugs and improve stability.

# Version 0.7.8 (2017-05-22)

* Fix many bugs and improve stability.

# Version 0.7.7 (2017-05-18)

* Fix many bugs and improve stability.

# Version 0.7.6 (2017-05-16)

* Change sort order of the candidates to more better
* Support maven parent pom
* Add server error logger
* Change default formatter to google-java-format
* Support make import from eclipse code format settings (meghanadaFormatter.xml)
* Fix many bugs and improve stability.

# Version 0.7.5 (2017-04-27)

* Exclude anonymous class from candidates
* Use maven ModelBuilder

# Version 0.7.4 (2017-04-13)

* Fix import all and optimize import bugs
* Fix jump to enum declaration bugs

# Version 0.7.3 (2017-04-03)

* Fix optimize import
* Fix el-doc API bugs

# Version 0.7.2 (2017-03-28)

* Fix NPE #11
* Improve code fomatter

# Version 0.7.1 (2017-03-23)

* Fix jump to declaration bugs

# Version 0.7.0 (2017-03-18)

## Highlights

* Support show declaration (eldoc support)
* Using kryo 4 and improve cache performance
* Improve gradle sub module build performance
* Fix zip and file stream memory leaks
* Fix broken android project completions
* Fix many bugs.


server/src/main/java/meghanada/Main.java
package meghanada;

import static java.util.Objects.isNull;

import io.opencensus.contrib.zpages.ZPageHandlers;
import io.opencensus.trace.Tracing;
import io.opencensus.trace.config.TraceConfig;
import io.opencensus.trace.samplers.Samplers;
import java.io.File;
import java.io.IOException;
import java.net.Socket;
import java.util.Objects;
import meghanada.config.Config;
import meghanada.server.Server;
import meghanada.server.emacs.EmacsServer;
import meghanada.system.Executor;
import meghanada.telemetry.TelemetryUtils;
import meghanada.utils.ClassPathUtils;
import meghanada.utils.FileUtils;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.HelpFormatter;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.logging.log4j.Level;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.apache.logging.log4j.core.LoggerContext;
import org.apache.logging.log4j.core.appender.FileAppender;
import org.apache.logging.log4j.core.config.Configuration;
import org.apache.logging.log4j.core.config.LoggerConfig;
import org.apache.logging.log4j.core.layout.PatternLayout;

public class Main {

  public static final String VERSION = "1.3.0";
  private static final Logger log = LogManager.getLogger(Main.class);
  private static String version;

  public static String getVersion() throws IOException {
    if (version != null) {
      return version;
    }
    final String v = FileUtils.getVersionInfo();
    if (v != null) {
      version = v;
      return version;
    }
    version = Main.VERSION;
    return version;
  }

  public static void main(String[] args) throws Exception {
    log.info("use java home: {}", System.getProperty("java.home"));
    log.info(
        "use java runtime: {} {}",
        System.getProperty("java.runtime.name"),
        System.getProperty("java.runtime.version"));
    log.info(
        "use java vm: {} {}",
        System.getProperty("java.vm.name"),
        System.getProperty("java.vm.version"));
    int size = Executor.getThreadPoolSize();
    System.setProperty(
        "java.util.concurrent.ForkJoinPool.common.parallelism", Integer.toString(size));
    ClassPathUtils.addToolsJar();
    final String version = getVersion();
    System.setProperty("meghanada-server.version", version);
    System.setProperty("meghanada-server.uid", TelemetryUtils.getUID());
    setupTelemetry();
    final Options options = buildOptions();

    final CommandLineParser parser = new DefaultParser();
    final CommandLine cmd = parser.parse(options, args);

    if (cmd.hasOption("h")) {
      HelpFormatter formatter = new HelpFormatter();
      formatter.printHelp("meghanada server : " + version, options);
      return;
    }
    if (cmd.hasOption("version")) {
      System.out.println(version);
      return;
    }
    Runtime.getRuntime()
        .addShutdownHook(
            new Thread(
                () -> {
                  Tracing.getExportComponent().shutdown();
                  log.info("shutdown server");
                  Config.showMemory();
                }));

    System.setProperty("home", Config.getInstalledPath().getParentFile().getCanonicalPath());

    addFileAppender(
        cmd.getOptionValue("log", System.getProperty("java.io.tmpdir") + "/meghanada_server.log"));

    if (cmd.hasOption("v")) {
      Object ctx = LogManager.getContext(false);
      if (ctx instanceof LoggerContext) {
        try (LoggerContext context = (LoggerContext) ctx) {
          Configuration configuration = context.getConfiguration();
          LoggerConfig loggerConfig = configuration.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
          loggerConfig.setLevel(Level.DEBUG);
          context.updateLoggers();
        }
      }
      System.setProperty("log-level", "DEBUG");
      log.debug("set verbose flag(DEBUG)");
    }

    if (cmd.hasOption("vv")) {
      Object ctx = LogManager.getContext(false);
      if (ctx instanceof LoggerContext) {
        try (LoggerContext context = (LoggerContext) ctx) {
          Configuration configuration = context.getConfiguration();
          LoggerConfig loggerConfig = configuration.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
          loggerConfig.setLevel(Level.TRACE);
          context.updateLoggers();
        }
      }
      System.setProperty("log-level", "TRACE");
      log.debug("set verbose flag(TRACE)");
    }

    if (cmd.hasOption("clear-cache")) {
      System.setProperty("clear-cache-on-start", "true");
    }

    if (cmd.hasOption("gradle-version")) {
      final String gradleVersion = cmd.getOptionValue("gradle-version", "");
      if (!version.isEmpty()) {
        System.setProperty("meghanada.gradle-version", gradleVersion);
      }
    }

    String port = "0";
    String projectRoot = "./";
    String fmt = "sexp";

    if (cmd.hasOption("p")) {
      port = cmd.getOptionValue("p", port);
    }
    if (cmd.hasOption("r")) {
      projectRoot = cmd.getOptionValue("r", projectRoot);
    }
    if (cmd.hasOption("output")) {
      fmt = cmd.getOptionValue("output", fmt);
    }
    log.debug("set port:{}, projectRoot:{}, output:{}", port, projectRoot, fmt);
    final int portInt = Integer.parseInt(port);

    log.info("Meghanada-Server Version:{}", version);
    final Server server = createServer("localhost", portInt, projectRoot, fmt);
    server.startServer();
  }

  private static void setupTelemetry() throws IOException {
    // global
    boolean exporter = TelemetryUtils.setupExporter();
    if (!exporter) {
      TraceConfig traceConfig = Tracing.getTraceConfig();
      traceConfig.updateActiveTraceParams(
          traceConfig
              .getActiveTraceParams()
              .toBuilder()
              .setSampler(Samplers.alwaysSample())
              .build());
      String zPort = System.getProperty("meghanada.zpage.port");
      if (isNull(zPort)) {
        int zpagePort = getFreePort();
        ZPageHandlers.startHttpServerAndRegisterAll(zpagePort);
        System.setProperty("meghanada.zpage.port", Integer.toString(zpagePort));
      } else {
        int zpagePort = Integer.parseInt(zPort);
        ZPageHandlers.startHttpServerAndRegisterAll(zpagePort);
      }
      System.setProperty("meghanada.zpage.enable", "true");
    }
  }

  private static void addFileAppender(String logFilename) throws IOException {
    File logFile = new File(logFilename);
    Object ctx = LogManager.getContext(false);
    if (ctx instanceof LoggerContext) {
      try (LoggerContext context = (LoggerContext) ctx) {
        Configuration configuration = context.getConfiguration();
        LoggerConfig loggerConfig = configuration.getLoggerConfig(LogManager.ROOT_LOGGER_NAME);
        FileAppender fileAppender =
            FileAppender.newBuilder()
                .setName("file")
                .setLayout(
                    PatternLayout.newBuilder()
                        .withPattern("[%d][%-5.-5p][%-14.-14c{1}:%4L] %-22.-22M - %m%n")
                        .build())
                .withFileName(logFile.getCanonicalPath())
                .build();
        configuration.addAppender(fileAppender);
        loggerConfig.addAppender(fileAppender, Level.ERROR, null);
        context.updateLoggers();
      }
    }
  }

  private static Server createServer(
      final String host, final int port, final String projectRoot, final String fmt)
      throws IOException {
    return new EmacsServer(host, port, projectRoot);
  }

  private static Options buildOptions() {
    final Options options = new Options();
    final Option help = new Option("h", "help", false, "show help");
    options.addOption(help);
    final Option version = new Option(null, "version", false, "show version information");
    options.addOption(version);
    final Option port = new Option("p", "port", true, "set server port. default: 55555");
    options.addOption(port);
    final Option project =
        new Option("r", "project", true, "set project root path. default: current path ");
    options.addOption(project);
    final Option verbose = new Option("v", "verbose", false, "show verbose message (DEBUG)");
    options.addOption(verbose);
    final Option traceVerbose =
        new Option("vv", "traceVerbose", false, "show verbose message (TRACE)");
    options.addOption(traceVerbose);
    final Option out =
        new Option(null, "output", true, "output format (sexp, csv, json). default: sexp");
    options.addOption(out);
    final Option gradleVersion = new Option(null, "gradle-version", true, "set use gradle version");
    options.addOption(gradleVersion);
    final Option clearCache = new Option("c", "clear-cache", false, "clear cache on start");
    options.addOption(clearCache);
    final Option logFilename =
        new Option("l", "log", true, "log file location. default: /tmp/meghanada_server.log");
    options.addOption(logFilename);
    return options;
  }

  public static boolean isDevelop() {
    final String dev = System.getenv("MEGHANADA_DEVELOP");
    return Objects.equals(dev, "1");
  }

  private static int getFreePort() throws IOException {
    int port;
    try (Socket socket = new Socket()) {
      socket.bind(null);
      port = socket.getLocalPort();
    }
    return port;
  }
}
